File: leetcode/C++/jump-game-ii.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        vector<int> visit(nums.size(),0);
        queue<int> q; q.push(0);
        if(nums.size()==1) return 0;
        for(int i=0,t; i<nums.size(); ++i){
            t=q.front(); q.pop();
            for(int j=1; j<=nums[t]; ++j) {
                if(j+t+1>=nums.size()) return visit[t] + 1;
                else if(visit[j+t]==0) {
                    q.push(j+t);
                    visit[j+t]=visit[t]+1;
                }
            }
        }
        return 0;
    }
};

File: leetcode/Python/jump-game-ii.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        visit = [0 for i in range(len(nums))]
        q, t = [0],0
        if len(nums) == 1:
            return 0
        for i in range(len(nums)):
            t = q.pop(0)
            for j in range(1,nums[t]+1):
                if j+t+1 >= len(nums):
                    return visit[t] + 1
                if visit[j+t] == 0:
                    q.append(j+t)
                    visit[j+t]=visit[t]+1
        return 0
'''
class Solution:
    def jump(self, nums: List[int]) -> int:
        sol, dist, cur = 1, 0, nums[0]
        if len(nums) == 1:
            return 0
        for i in range(len(nums)):
            if i+nums[i] > dist:
                dist = i+nums[i]
            if i == cur:
                cur = dist
                sol += 1
            if cur+1 >= len(nums):
                return sol
        return 0
'''